llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,169 @@
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+#include <set>
+#include <unordered_map>
+
+#define DEBUG_TYPE "aie-duplicate-PHI-elimination"
+
+using namespace llvm;
+
+static const char AIE_ELIMINATE_DUPLICATE_PHI[] =
+    "AIE Eliminate Duplicate PHI Pass";
+
+namespace {
+
+class AIEEliminateDuplicatePHI : public MachineFunctionPass {
+public:
+  static char ID;
+  const int DefaultAddrSpace = 0;
+  AIEEliminateDuplicatePHI() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE to set up GISelObserverWrapper object
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    bool Changed = false;
+    for (MachineBasicBlock &MBB : MF) {
+      // If a BB has PHI nodes, the very first instruction will be PHI
+      if (MBB.empty() ||
+          MBB.begin()->getOpcode() != llvm::TargetOpcode::G_PHI) {
+        continue;
+      }
+      Changed = processMBBWithPHI(MBB, MRI, MIB, Observer);
+    }
+    return Changed;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addRequired<GISelCSEAnalysisWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
+    AU.setPreservesAll();
+  }
+
+  StringRef getPassName() const override { return AIE_ELIMINATE_DUPLICATE_PHI; }
+
+private:
+  bool eliminateDuplicatePHI(
+      std::set<std::pair<MachineInstr *, MachineInstr *>> &IdenticalPHI,
+      MachineIRBuilder &MIB, GISelObserverWrapper &Observer) {
+    bool Changed = false;
+    if (!IdenticalPHI.empty()) {
+      LLVM_DEBUG(dbgs() << "Identical PHI nodes found\n");
+      for (auto &[BasePHI, DuplicatePHI] : IdenticalPHI) {
+        LLVM_DEBUG(dbgs() << "Base PHI: " << *BasePHI << "\n");
+        LLVM_DEBUG(dbgs() << "Duplicate PHI: " << *DuplicatePHI << "\n");
+        MIB.setInstr(*DuplicatePHI);
+        MIB.buildCopy(DuplicatePHI->getOperand(0).getReg(),
```

Comment:
Can't we change the uses instead of creating a COPY?

Reply:
Not creating any new instructions and replacing the use of duplicate PHI